src/store.rs

```diff
@@ -15,6 +15,21 @@ pub trait Storage: Debug {
     fn add_contract(&mut self, hash: U256, code: Vec<U256>) -> Result<(), StorageError>;
     fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError>;
     fn storage_write(&mut self, key: StorageKey, value: U256) -> Result<(), StorageError>;
+    fn get_all_keys(&self) -> Vec<StorageKey>;
+    fn fake_clone(&self) -> InMemory;
+    fn rollback(&mut self, previous: &dyn Storage) {
+        let keys = previous.get_all_keys();
+        for key in keys {
+            let value = previous.storage_read(key).unwrap().unwrap();
+            self.storage_write(key, value).unwrap();
+        }
+        let current_keys = self.get_all_keys();
+        for key in current_keys {
+            if previous.storage_read(key).is_err() {
+                self.storage_write(key, U256::zero()).unwrap();
+            }
+        }
+    }
```

Comment:
I believe it makes more sense to rollback by default, by just dropping the frame, and have a consuming method to apply to the previous state instead. I would also suggest using an `IntoIterator` implementation rather than creating a (potentially huge) vector of keys. The proposed interfaces would be:
```rust
fn into_iter(self) -> impl Iterator<Item = (StorageKey, U256)>; // Or rather, require this as a trait bound
fn apply(self, target: &dyn mut Storage) { // Only called on successful return
    for (k, v) in self.into_iter() { // Can omit the `.into_iter()` if `Self` implements `IntoIterator`
        target.storage_write(k, v);
    }
}
```

Reply:
That is a nice approach, I left this PR on draft since I'm currently doing other work, the idea is that doing rollbacks fixes this tests, but It is necessary to think of a better way of performing rollbacks, this could be an option.
We still have the issue of cloning the Storage when implemented with rocksdb, that is another key part to think.

Reply:
We may also avoid the need to `.fake_clone()` simply by making `storage_read()` implementations recursively ask the previous links in the chain. The `struct`s may simply store an optional reference to the previous frame's storage. The number of queries will be bound to the depth of the call stack.